### ISSUE TITLE
Undefined names

### ISSUE THREAD

**Issue by cclauss** [kengz/SLM-Lab#135]
Undefined names have the potential to raise NameError at runtime.

flake8 testing of https://github.com/kengz/SLM-Lab on Python 3.6.3

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./slm_lab/agent/algorithm/base.py:73:16: F821 undefined name 'action'
        return action
               ^
./slm_lab/agent/algorithm/base.py:99:16: F821 undefined name 'batch'
        return batch
               ^
./slm_lab/agent/algorithm/policy_util.py:43:13: F821 undefined name 'new_prob'
            new_prob[torch.argmax(probs, dim=0)] = 1.0
            ^
./slm_lab/env/__init__.py:97:49: F821 undefined name 'nvec'
        setattr(gym_space, 'low', np.zeros_like(nvec))
                                                ^
./slm_lab/experiment/search.py:131:9: F821 undefined name 'config'
        config['trial_index'] = self.experiment.info_space.tick('trial')['trial']
        ^
./slm_lab/experiment/search.py:133:16: F821 undefined name 'config'
        return config
               ^
./slm_lab/experiment/search.py:146:16: F821 undefined name 'trial_data_dict'
        return trial_data_dict
               ^
./test/agent/net/test_nn.py:83:25: F821 undefined name 'net_util'
        before_params = net_util.copy_trainable_params(net)
                        ^
./test/agent/net/test_nn.py:88:24: F821 undefined name 'net_util'
        after_params = net_util.copy_trainable_params(net)
                       ^
./test/agent/net/test_nn.py:114:25: F821 undefined name 'net_util'
        before_params = net_util.copy_fixed_params(net)
                        ^
./test/agent/net/test_nn.py:118:24: F821 undefined name 'net_util'
        after_params = net_util.copy_fixed_params(net)
                       ^
11    F821 undefined name 'action'
11
```

**Comment by cclauss** [comment]
See #136 [flake8](http://flake8.pycqa.org) will find _undefined names_ which might not be discovered by other linters or tests.

Each undefined name has the potential to raise NameError at runtime.

**Comment by kengz** [comment]
unclear what problem this aims to solve; there is no breakage in the code referred. closing as irrelevant.

**Comment by kengz** [comment]
reopening issue:
TODO: see #137 for some context; write proper regression tests.

**Comment by kengz** [comment]
fixed by #139

**Comment by cclauss** [comment]
From 11 above down to 7 below...  _Undefined names_ have the potential to raise _NameError_ at runtime.

[flake8](http://flake8.pycqa.org) testing of https://github.com/kengz/SLM-Lab on Python 3.7.0

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./slm_lab/env/base.py:25:49: F821 undefined name 'nvec'
        setattr(gym_space, 'low', np.zeros_like(nvec))
                                                ^
./slm_lab/agent/algorithm/base.py:84:16: F821 undefined name 'action'
        return action
               ^
./slm_lab/agent/algorithm/base.py:91:16: F821 undefined name 'batch'
        return batch
               ^
./slm_lab/lib/viz.py:299:16: F821 undefined name 'TimeoutExpired'
        except TimeoutExpired:
               ^
./slm_lab/experiment/search.py:133:9: F821 undefined name 'config'
        config['trial_index'] = self.experiment.info_space.tick('trial')['trial']
        ^
./slm_lab/experiment/search.py:135:16: F821 undefined name 'config'
        return config
               ^
./slm_lab/experiment/search.py:148:16: F821 undefined name 'trial_data_dict'
        return trial_data_dict
               ^
7     F821 undefined name 'action'
7
```
